frontend/src/main/scala/bloop/engine/tasks/CompileTasks.scala

```diff
@@ -0,0 +1,117 @@
+package bloop.engine.tasks
+
+import bloop.engine.Dag
+import bloop.{CompileInputs, Compiler, Project}
+
+import scala.concurrent.duration.Duration
+import scala.concurrent.Await
+import xsbti.compile.{CompileAnalysis, MiniSetup, PreviousResult}
+import bloop.reporter.{Reporter, ReporterConfig}
+import sbt.internal.inc.{ConcreteAnalysisContents, FileAnalysisStore}
+
+object CompileTasks {
+  import bloop.engine.State
+  def compile(state: State, project: Project, reporterConfig: ReporterConfig): State = {
+    import State.compilerCache
+    import state.logger
+    def toInputs(project: Project, config: ReporterConfig, result: PreviousResult) = {
+      val instance = project.scalaInstance
+      val sourceDirs = project.sourceDirectories
+      val classpath = project.classpath
+      val classesDir = project.classesDir
+      val target = project.tmp
+      val scalacOptions = project.scalacOptions
+      val javacOptions = project.javacOptions
+      val reporter = new Reporter(logger, project.baseDirectory.syntax, identity, config)
+      // FORMAT: OFF
+      CompileInputs(instance, compilerCache, sourceDirs, classpath, classesDir, target, scalacOptions, javacOptions, result, reporter, logger)
+      // FORMAT: ON
+    }
+
+    val results = new java.util.concurrent.ConcurrentHashMap[Project, PreviousResult]()
+    def runCompilation(project: Project): Seq[Project] = {
+      val previousResult = state.results
+        .getResult(project)
+        .getOrElse(sys.error("Results cache was not initialized"))
+      val inputs = toInputs(project, reporterConfig, previousResult)
+      val result = Compiler.compile(inputs)
+      results.put(project, result)
+      List(project)
+    }
+
+    def compileTask(project: Project) =
+      new Task((_: Seq[Project]) => runCompilation(project), () => ())(Mergeable.SeqMergeable)
```

Comment:
The implicit is not found otherwise?

Reply:
Could be found, this is just redundant. I'll change this to a map anyway to address your other comment.